jgrapht-core/src/main/java/org/jgrapht/graph/AsSynchronizedGraph.java

```diff
@@ -0,0 +1,1114 @@
+    {
+        return c.getClass() == CacheAccess.class;
+    }
+
+    /**
+     * Set the cache strategy for <code>edgesOf</code>, <code>incomingEdgesOf</code> and
+     * <code>outgoingEdgesOf</code> methods.
+     * @param c the cache strategy, if <tt>true</tt>, cache will be used for those methods, otherwise
+     *          cache will not be used.
+     */
+    public void setCache(boolean c)
+    {
+        if (c == useCache())
+            return;
+        if (c)
+            this.c = new CacheAccess();
+        else
+            this.c = new NoCache();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int hashCode()
+    {
+        synchronized (mutex) {
+            return delegate.hashCode();
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o)
+            return true;
+        synchronized (this) {
+            return delegate.equals(o);
+        }
+    }
+
+
+    /**
+     * Create a copy of the set.
+     *
+     * @param set the set to be copied.
+     *
+     * @return a copy of the set
+     */
+    private <C> Set<C> copySet(Set<C> set)
+    {
+        return Collections.unmodifiableSet(new LinkedHashSet<>(set));
+    }
+
+    /**
+     * Inform allVerticesSet that the backing data has be modified.
+     */
+    private void vertexSetModified()
+    {
+        if (allVerticesSet != null)
+            allVerticesSet.modified();
+    }
+
+    /**
+     * Inform allEdgesSet that the backing data has be modified.
+     */
+    private void edgeSetModified()
+    {
+        if (allEdgesSet != null)
+            allEdgesSet.modified();
+    }
+
+    /**
+     * Create a synchronized (thread-safe) and unmodifiable Set backed by the specified Set. In order
+     * to guarantee serial access, it is critical that <strong>all</strong> access to the backing
+     * Set is accomplished through the created Set.
+     *
+     * <p>
+     * When users get this Set's <code>Iterator</code>, <code>Spliterator</code>, <code>Stream</code>
+     * or <code>ParallelStream</code>, this Set will copy itself and return the copy's
+     * <code>Iterator</code>, <code>Spliterator</code>, <code>Stream</code> or <code>ParallelStream</code>.
+     * If the backing Set is modified. Users need to call <code>modified</code> method to inform the
+     * Set that the copy need to update.
+     * </p>
+     *
+     * <p>
+     * The created Set's hashCode is equal to the backing Set's hashCode. And the created Set is equal
+     * to another set if they are the same Set or the backing Set is equal to the other Set.
+     * </p>
+     *
+     * <p>
+     * The created set will be serializable if the backing set is serializable.
+     * </p>
+     *
+     *
+     * @param <E> the class of the objects in the set
+     *
+     * @author CHEN Kui
+     * @since Feb 23, 2018
+     */
+
+    private class AsSynchronizedIterateOnCopySet<E>
```

Comment:
Is there any need for this inner class to be non-static?  Also, since it's already part of a class named AsSynchronizedGraph, a shorter name such as "CopyOnDemandSet" will do.

Reply:
I think the set will only be used for to synchronized graph.

Reply:
Right, using an inner class is fine, but there's a difference between a static inner class and a non-static inner class.